service/worker/addsearchattributes/workflow.go

```diff
@@ -0,0 +1,203 @@
+	}
+
+	updateClusterMetadataActivityOptions = workflow.ActivityOptions{
+		RetryPolicy: &temporal.RetryPolicy{
+			InitialInterval: 1 * time.Second,
+		},
+		StartToCloseTimeout:    2 * time.Second,
+		ScheduleToCloseTimeout: 10 * time.Second,
+	}
+
+	ErrUnableToUpdateESMapping      = errors.New("unable to update Elasticsearch mapping")
+	ErrUnableToExecuteActivity      = errors.New("unable to execute activity")
+	ErrUnableToGetSearchAttributes  = errors.New("unable to get search attributes from cluster metadata")
+	ErrUnableToSaveSearchAttributes = errors.New("unable to save search attributes to cluster metadata")
+)
+
+func newActivities(
+	esClient esclient.Client,
+	saManager searchattribute.Manager,
+	metricsClient metrics.Client,
+	logger log.Logger,
+) *activities {
+	return &activities{
+		esClient:      esClient,
+		saManager:     saManager,
+		metricsClient: metricsClient,
+		logger:        logger,
+	}
+}
+
+// AddSearchAttributesWorkflow is the workflow that
+func AddSearchAttributesWorkflow(ctx workflow.Context, params WorkflowParams) error {
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Workflow started.", "wf-type", WorkflowName)
+
+	var a *activities
+
+	ctx1 := workflow.WithActivityOptions(ctx, addESMappingFieldActivityOptions)
+	err := workflow.ExecuteActivity(ctx1, a.AddESMappingFieldActivity, params).Get(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("%w: AddESMappingFieldActivity: %v", ErrUnableToExecuteActivity, err)
+	}
+
+	ctx2 := workflow.WithActivityOptions(ctx, waitForYellowStatusActivityOptions)
+	err = workflow.ExecuteActivity(ctx2, a.WaitForYellowStatusActivity, params.IndexName).Get(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("%w: WaitForYellowStatusActivity: %v", ErrUnableToExecuteActivity, err)
+	}
+
+	ctx3 := workflow.WithActivityOptions(ctx, updateClusterMetadataActivityOptions)
+	err = workflow.ExecuteActivity(ctx3, a.UpdateClusterMetadataActivity, params).Get(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("%w: UpdateClusterMetadataActivity: %v", ErrUnableToExecuteActivity, err)
+	}
+
+	// TODO: anything better?
+	workflow.GetMetricsScope(ctx).SubScope("AddSearchAttributesWorkflow").Counter("add_search_attributes_workflow_success").Inc(1)
+	logger.Info("Workflow finished successfully.", "wf-type", WorkflowName)
+	return nil
+}
+
+func (a *activities) AddESMappingFieldActivity(ctx context.Context, params WorkflowParams) error {
+	mapping := make(map[string]string, len(params.CustomAttributesToAdd))
+	for saName, saType := range params.CustomAttributesToAdd {
+		esType := searchattribute.MapESType(saType)
+		if esType == "" {
+			return temporal.NewNonRetryableApplicationError(fmt.Sprintf("Unknown search attribute type: %v", saType), "", nil)
+		}
+		mapping[saName] = esType
+	}
+
+	a.logger.Info("Creating Elasticsearch mapping.", tag.ESIndex(params.IndexName), tag.ESMapping(mapping))
+	_, err := a.esClient.PutMapping(ctx, params.IndexName, searchattribute.Attr, mapping)
```

Comment:
is it worth to call GetMapping before calling PutMapping?
or is PutMapping idempotent?

Reply:
`mapping` has only new fields and yes, `PUT` is idempotent (I tested this specific case).